### DISCUSSION TITLE
VLC connects only when on local network

### DISCUSSION THREAD

**Discussion by remgr** [GerryDazoo/Slinger#105] [Q&A] [answered]
First and foremost, thank you Gerry.  Between my brother and I we have 5 Slingboxes that you have saved.  I am only moderately tech savvy, and I was able to setup the boxes, and each runs successfully on the pc and android version of VLC when I am located on my home network,  When I try to connect remotely,  I am unsuccessful, and VLC reports that it can't connect.  I am not sure of what troubleshooting steps I should take to correct this.  Please advise.

**Comment by GerryDazoo** [comment, marked as answer]
When you say “remotely” do you mean from the Internet or another box on your local LAN?

1.	Make sure your local firewall where the server is running is allowing connections to [SERVER] port=xxx
2.	To allow remote access from WAN add the appropriate port forwarding rule to your router.

**Reply by remgr** [reply]
Gerry, thank you for your quick response.  I have spent hours today, trying various firewall settings.  I believe that I did it correctly, I am more a pc person then a network person.  I still can only connect on the router where I am running the server.exe.  I am attaching a picture to show you what is happening.  The left side is the home PC, with VLC running.  The right side is my office PC (throught Chrome Remote Desktop).  When I try to access the slingbox I get an error message.  I have enabled WAN access on my Asus router, and even set up a rule for the Slingbox ip address.  Not sure what I'm doing wrong.

Another issue I have is in the office we have multiple slingboxes, and the server.exe asks to select the slingbox I want.  Is there a line I should uncomment and edit to address this?  I am including a screenshot of that too.  Thanks

![Error Screen](https://user-images.githubusercontent.com/114819659/193469991-95f3db8f-68d3-4f6d-8358-8b3c5b3b447e.png)

![Error Multiple Slingboxes](https://user-images.githubusercontent.com/114819659/193470016-aee7e008-b0da-49e3-9655-749f05f2ea38.png)

**Comment by GerryDazoo** [comment]
You have many problems.
First your not giving the server a valid IP address for your slingbox. ( See connect error line). When it can't connect it tries to scan your local network for slingboxes and it can't find any so it gives up. You need to have a port forwarding rule on the router at 96.250.82.132... to get packets to your slingbox.
I don't know why in one case it's finds a M1 on the network and uses it but in the second case it doesn't. 
Also, please don't send screenshots my old eyes have a problem seeing them and sometimes useful information scrolls off the top of the window. Cut and paste the text and post that. In this case the vlc window is obscuring valuable information.
For now turn off your local firewall and fix the port forwarding rules and retest.

**Comment by remgr** [comment]
Gerry hello.  I am pulling my hair out trying to figure out where my mistake is.  These are the steps I have taken.  I moved one of the slingboxes to another router to avoid all the conflicts.  I opened a port in the new (Netgear) router with the external port set to 5316, and the internal port at 5316.  This is the port Slingerplayer shows on network setting.    I set the slingbox ip to static.  I allowed VLC & the server.exe through the Windows firewall, checking both.  Once again it works fine when I use the local ip address on VLc, but when I use the external address, on VLC network stream, with the port number I opened,  it does not work.  I have attached the VLC log as well as the Netgear log to show you what they report.  I know I am missing something, but can't figure out what.

[VLC Error Log.txt](https://github.com/GerryDazoo/Slinger/files/9719274/VLC.Error.Log.txt)

[Netgear Error.txt](https://github.com/GerryDazoo/Slinger/files/9719279/Netgear.Error.txt)

**Reply by GerryDazoo** [reply]
Unfortunately, I can't help you unless you supply the information I asked for. I need the server output. The URLs you are supplying to VLC and the contents for your config.ini files. All I could gleen from your message is the slingbox's local IP is 192.168.1.70 
Again, are you trying to use the slingbox remotely, or the server? If it's the slingbox then you do NOT give VLC the the external IP address but the local server IP address and you configure my code with the Slingbox's external IP address.  It's impossible to tell what your trying to do because it looks like your using the same port number for the server and the slingbox.  I believe I recommended earlier NOT to do that because it is certainly causing me confusion.
Are you trying to access the server remotely? If so the port forwarding rule needs to point to the IP address where the server is running NOT the slingbox. 

So, when you connect locally you'll see the sever generate some logs saying it's starting up a stream. If you don't see those logs when you try to connect to the server remotely then either your port forwarding rule is missing or wrong or your local firewall on the host where the server is running is blocking the connections. There's not much more to it than that and is basically what I said in my first response.

**Comment by remgr** [comment]
Hi Gerry, again thank you for the help.  What I want to do is view the slingbox remotely on VLC.   I am running the server software at home, but the slingbox is physically located in the office.  When I use my home pc to try to view the office slingbox on the office ip 71.105... it does not work on VLC.  When I try to view it in my home on my pc, with my ip address 192.168..., the office slingbox shows perfectly.    

To be specific, I have this address on VLC : http://192.168.50.187:5322/slingbox which works immediately on my desktop.  I then try this address: http://71.105.100.26:5316/slingbox, and port 5322 and port 5323 (I tried all of them).  I get this error message from VLC: "Connection refused by peer"

I have also tried running the server on my office pc, but again VLC will only show up on the local 192.168... ip of the office router.

I have allowed both VLC and the slingbox_server for public and private on Windows Firewall.  I have disabled my antivirus as well.

I know pictures are difficult to decipher, but I can't copy the text from the slingbox_server cmd screen.   I attached a screen shot of the server I am running now, and the port forwarding port that I have set up on the router in the office.   The slingbox is set as a static ip address as well.  I am also attaching the congif.ini file converted to text.

Is it the case that I should do something on my router here at home?

  
[config.txt](https://github.com/GerryDazoo/Slinger/files/9728988/config.txt)
![Server Screen](https://user-images.githubusercontent.com/114819659/194432809-d655df8a-e1dc-453c-9704-d35c4e582141.jpg)
![Netgear](https://user-images.githubusercontent.com/114819659/194432819-abee8c55-5448-48e4-9f23-2788e3613256.jpg)

**Reply by GerryDazoo** [reply]
OK. You need to point VLC to where the SLingBox Server is running NOT the slingbox.
http://71.105.100.26:5316/slingbox  This is Wrong. This is the slingbox not my server code.
If your running the server on your home PC and trying to connect from outside your house then you need a port forwarding rule on your home router to get packets to the server
port 5322 internal ip 192.168.50.187

So to view the office slingbox from anywhere in the world us this URL for VLC.  http://YourHomeExternalIpAddress:5322/slingbox

Please spend a few minutes googling how to cut and paste text from a command prompt window.

**Comment by remgr** [comment]
Hi Gerry,

Success!  Thank you for all your help.  I have managed to get everything to work successfully on one router at work.  I can view at home, on another pc through Google Remote Desktop, and VLC on my phone! There are still issues I am working to resolve on others.  I will invest the time needed to work everything else out.  

I do seem to have a hit and miss problem running the slingbox-server from my home pc.  I get the error message:
Exception in thread Thread-2 (streamer)
Traceback (most recent call last):
   File "treading .py" line 1016, in _bootstrap inner
   File "treading .py" line 953, in run
   File "sling_server.py" line 410 in streamer
     for port in stati.keys:
TypeError: 'bulitin_function_or_method'object is not iterable

I can then go back a few minutes later, and the server runs correctly.  Would setting a static ip for my home pc help?  

As I said, I do have a working solution.  I will now have to tackle this from my brother's pc, he has several slingboxes himself.

You have been most kind, and I do appreciate your help.  I am a little old lady myself, 6 months away from retirement!

**Reply by GerryDazoo** [reply]
This is a bug in the server code. Please make sure you have a valid slingbox IP address and port number configured in your config.ini.
Looks like you figured out how to cut and paste the program output. But, you still didn't supply all the program output.

**Comment by remgr** [comment]
Hi Gerry,

Thank you for your quick response.  I have spent hours today, trying
various firewall settings.  I believe that I did it correctly, I am more a
pc person then a network person.  I still can only connect on the router
where I am running the server.exe.  I am attaching a picture to show you
what is happening.  The left side is the home PC, with VLC running.  The
right side is my office PC.  When I try to access the slingbox I get an
error message.  I have enabled WAN access on my Asus router, and even set
up a rule for the Slingbox ip address.  Not sure what I'm doing wrong.

Another issue I have is in the office we have multiple slingboxes, and the
server.exe asks to select the slingbox I want.  Is there a line I should
uncomment and edit to address this?  I am including a screenshot of that
too.  Thanks

**Reply by GerryDazoo** [reply]
You are missing something very important. For every different slingbox you want the server to connect to you need a different config.ini file with the appropriate slingbox IP and password  and video configuration information and a unique server port. So if you have 3 slingboxes you would give the server the three config.ini file names on the command line.  
slingbox_server.exe config1.ini config2.ini config3.ini 
The [SERVER] ports could be 8080, 8082 and 8084 or anything you might like to choose. I know the multiple slingboxes setup is complicated but it does work, I've got 4 boxes configured for my testing at home. The next version of the server will make this much easier but in the meantime, that's what you need to do.
When the server can't connect to the configured slingbox, it scans the local network looking for boxes. This is what is happening to you. You must make sure the IP info for the slingbox is correct and if it's a remote slingbox that there is the appropriate port forwarding rule applied to the router in front of the slingbox NOT the router where the server is running.
How many slingboxes do you have at the office?
PS. If you continue to refuse to supply the diagnostic information I've been continually asking for, server output, config.ini information and URLs you are using to connect to the server I'm not going to be able to help you further.

**Comment by remgr** [comment]
Hi Gerry - I actually had folders and different config files for each Slingbox.  My mistake was moving one of them to the second ISP we have in the office (dedicated to the voip phones).  That router for the phones is much more user friendly. I had scrambled to activate all the spare slingboxes in my possession, in order to create the administrator password.  There were several. I still didn't have a feel for what exactly was happening with the slingbox_server.exe. I now understand the concept thanks to your previous answer.  I ran that config.ini without changing the public IP etc.  All are working correctly.

The only hurdle I have left is setting up my brother's slingboxes.  He has an Altice modem that does not allow static IPs on devices.  Creating a port forward is difficult.  The ip address (from ipconfig/all & the slingbox_server app) for the pc doesn't match the the router distribution range which is 192.168.1.11-254.  Ipconfig/all show it as 192.168.4.152.  Not sure how to handle that.  I am planning to set up a port trigger for port 5301.  Will that work?  Should I setup one for VLC on port 8080 as well?  I have attached that config file.
[config.txt](https://github.com/GerryDazoo/Slinger/files/9760202/config.txt)

Again, many thanks for your help.
